### DISCUSSION TITLE
Netherite "Exactly-Once" Guarantee Clarification

### DISCUSSION THREAD

**Discussion by gorillapower** [microsoft/durabletask-netherite#288] [Q&A]
Im interested in more detail around the execution guarantees that Netherite provides.

In the documentation
> Like the other back-ends for DF and DTFx, Netherite provides **at-least-once** guarantees for executing activities. It also provides some additional, stronger guarantees:

But also states

> All communication within a TaskHub (i.e. between orchestrations, activities, and entities) is transactionally consistent **(“exactly-once”)**, since the EventHubs API allows us to deliver messages between partitions without any risk of duplication.

Can anyone clarify what the difference between "communication within TaskHub" (providing 'exactly-once' delivery) and say the communication between an Orchestration and an Activity which provides "at-least-once" delivery?  Further, what constitutes communication within a TaskHub that allows it to be "exactly-once"?

We are using Netherite and consistently getting "at-least-once" behaviour on an Activity, that is being signalled from an Orchestrator within the same Taskhub, but I gather this is due to the the guarantee based on DF/DTFx framework rather than one that is to do with Netherite? I wondering if it is possible to achieve "exactly-once" for activities. 

P.s. We are also communicating between EventHubs, using the `IDurableEntityClient` but I assume this would break the 'exactly-once' guarantee since its across EventHubs.

**Comment by davidmrdavid** [comment]
cc/ @sebastianburckhardt
